feat(hardware): Added a driver to read and write data from the Flex eeprom

vegano1 opened the pull request:
# Overview

We need a way to read/write to the eeprom on the carrier board, this python module will let us do that from userspace. 

Closes: [RET-1346](https://opentrons.atlassian.net/browse/RET-1346)

**Property Format**
Data is written to the eeprom in the form of serialized property byte string with the following format
`prop_id(1b) + data_length(1b) + data(1-253b)`

- The `prop_id` is a 1-byte 1-0xFE value and denotes the unique property being stored: **NOTE: 0xFF is invalid**
- The `data_length` is a 1-byte 1-0xFD value and denotes the length of the data 
- The `data` is a 1-253b long hexadecimal value representing some data that corresponds to a prop_id.

**Properties**
Properties represent some piece of data stored on the eeprom, and are categorized by a `prop_id`, a `data_type`, and a `value`. When we find a valid serialized property byte string, which means we have a bytestring matching our format denoted above where the `length` byte corresponds to the actual size of the data. We deserialize this bytestring based on the specific prop_id, the type of data that prop_id is associated with, and how long the data should be. We use the data_type to encode/decode the data accordingly and create a Property object with the parsed data type.

**API**
In order to interact with the eeprom we have the concept of properties, and we use the methods `self.property_write` and `self.property_read` to read and write these properties to the eeprom.

The `self.property_write` method takes a set of property and value tuples `Set[Tuple[property_id, Any]])` which are serialized, chunked up, and written to the eeprom. This method returns a set of PropId for the properties that were successfully written to the eeprom.

The `self.property_read` method takes a set of PropId enums and returns a set of Property objects which represent the eeprom data.


# Test Plan

- [x] Make sure we can read/write data to/from the eeprom on the flex


# Changelog

- Added eeprom module that lets us read/write data to/from any eeprom device in `/sys/bus/i2c/devices`

# Review requests

- Does this approach make sense?
- Am I missing something?

# Risk assessment
Low, flex work

[RET-1346]: https://opentrons.atlassian.net/browse/RET-1346?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
eeprom module wip

vegano1 committed:
added ability to read property from eeprom

vegano1 committed:
switch read interface to Set instead of List
fix an issue when reading properties from the eeprom

vegano1 committed:
added raw serial number propId

vegano1 committed:
add property_write function and clean up

sfoster1 commented on hardware/opentrons_hardware/drivers/eeprom/types.py:
```diff
@@ -0,0 +1,71 @@
+    PropType.SHORT: 2,
+    PropType.INT: 4,
+    PropType.STR: 254,
+    PropType.BIN: 254,
```
same?

sfoster1 reviewed (changes requested):
I think if we're doing TLV, we should do TLV all the way. We can drop the start and end tokens, because we have the type, length and value. Require the type to not be 255 and you can't be confused with erased data; then, you can read out `size-1` bytes, nice and easy.

vegano1 committed:
Merge branch 'edge' into RET-1346_add_eeprom_module

vegano1 committed:
add gpio_dev to EEPROM object so we can set the eeprom write protect line SODIMM_222 low when writting to the eeprom

vegano1 committed:
remove swap files

vegano1 committed:
remove start/end tokens and implement TLV
deal with overflowing incomplete data
cleaning up

vegano1 committed:
- added filepath arg to EEPROM so we can override the default sys path for testing
- removed hard while loops when parsing data
- return set of PropertyId of the properties that were successfully written to the eeprom
- dropped the manufacturing_facility and unique_id values from the serial number
- removed unused PropertyId’s MACHINE_TYPE, MACHINE_VERSION, PROGRAMMED_DATE, UNIT_NUMBER
- added MAX_DATA_LEN so we can limit the size of a property to 255 bytes
- removed eeprom_writter.py script as this will be added in a separate pr
- added unit tests

vegano1 committed:
Merge branch 'edge' into RET-1346_add_eeprom_module

vegano1 committed:
created SystemDrivers holding class to make it easier for _build_system_hardware to return one item
added defines to drivers.__init__
use gpiod passed down from ot3Controller when initializing the EEPROMDriver

vegano1 committed:
change filepath to eeprom_path and make a Path type insted of string
return the EEPROMDriver module from the OT3Controller via a property
fix the eeprom_build_driver module

vegano1 committed:
Merge branch 'edge' into RET-1346_add_eeprom_module

vegano1 committed:
fix unit tests

vegano1 committed:
fixed an issue with data overflow
added unit tests

vegano1 committed:
_get_eeprom_info to get the name and size of the eeprom

sfoster1 commented on hardware/opentrons_hardware/drivers/eeprom/eeprom.py:
```diff
@@ -0,0 +1,252 @@
+
+        try:
+            logger.debug(
+                f"Writting {len(data)} bytes to address {hex(address)} for device {self.name} - {data.hex()}"
```
```suggestion
                f"Writing {len(data)} bytes to address {hex(address)} for device {self.name} - {data.hex()}"
```

sfoster1 reviewed (approved):
Looks good to me!

vegano1 committed:
fix typo
